pkg/iter/iterator.go

```diff
@@ -133,38 +135,67 @@ func (i *heapIterator) requeue(ei EntryIterator) {
 }
 
 func (i *heapIterator) Next() bool {
-	if i.curr != nil {
-		i.requeue(i.curr)
-	}
-
 	if i.heap.Len() == 0 {
 		return false
 	}
 
-	i.curr = heap.Pop(i.heap).(EntryIterator)
-	currEntry := i.curr.Entry()
-
-	// keep popping entries off if they match, to dedupe
+	// We support multiple entries with the same timestamp, and we want to
+	// preserve their original order. We look at all the top entries in the
+	// heap with the same timestamp, and pop the ones whose common value
+	// occurs most often.
+	type tuple struct {
+		logproto.Entry
+		EntryIterator
+	}
+	tuples := make([]tuple, 0, i.heap.Len())
 	for i.heap.Len() > 0 {
 		next := i.heap.Peek()
-		nextEntry := next.Entry()
-		if !currEntry.Equal(nextEntry) {
+		entry := next.Entry()
+		if len(tuples) > 0 && !tuples[0].Timestamp.Equal(entry.Timestamp) {
 			break
 		}
 
-		next = heap.Pop(i.heap).(EntryIterator)
-		i.requeue(next)
+		heap.Pop(i.heap)
+		tuples = append(tuples, tuple{
+			Entry:         entry,
+			EntryIterator: next,
+		})
+	}
+
+	// Find in entry which occurs most often which, due to quorum based
+	// replication, is guaranteed to be the correct next entry.
+	sort.Slice(tuples, func(i, j int) bool {
+		return tuples[i].Line < tuples[j].Line
+	})
+	i.currEntry = tuples[0].Entry
+	count, max := 1, 1
+	for j := 1; j < len(tuples); j++ {
+		if tuples[j].Equal(tuples[j-1]) {
+			count++
+			continue
+		}
+		if count > max {
+			i.currEntry = tuples[j-1].Entry
+			max = count
```

Comment:
I think you need to reset count to `0` here?

Reply:
Actually, we need a few other things here. If the tuples were "a", "b", "b"  then on the first iteration of the loop, tuple[j] wont match tuple[j-1], but both count and max will still be equal to 1, so i.currEntry wont be updated.  On the second and final iteration, tuple[j] will equal tuple[j-1], and the loop will exit. again leaving i.currEntry unchanged, even though we should be updating it.

```
	i.currEntry = tuples[0].Entry
	count, max := 0, 0
	for j := 0; j < len(tuples)-1; j++ {
		// if next equals current, increment our count of duplicates
		if tuples[j+1].Line == tuples[j].Line {
			count++
			continue
		}
		// there are no more duplicates for this line. If the number of
		// duplicates is greater then for any other seen line, use this entry
		if count > max {
			i.currEntry = tuples[j].Entry
			max = count
			count = 0
		}
	}

	// check if the last tuple is the one with the most duplicates
	if count > max {
		i.currEntry = tuples[len(tuples)-1].Entry
	}
```

Reply:
Yeah, missed the case where last could be most common.  Have factored out and fixed.